### DISCUSSION TITLE
Check for IntegrationEvents in internal objects (codeunits)

### DISCUSSION THREAD

**Discussion by fvet** [StefanMaron/BusinessCentral.LinterCop#144] [New Rule]
**Scenario**
We started applying rule **LC0011** on most of our objects (especially codeunits) and decided to put all on **Internal** in a first step.
However, lots of our codeunits use the pattern below (~ method codeunit pattern) 

```
Codeunit

Access = Internal <<< conflict : it blocks the IntegrationEvents from being used

Internal Function (called via it's table)
Local DoFunction
IntegrationEvent OnBefore <<<
IntegrationEvent OnAfter <<<
```

If we set the Access to Internal on codeunit level, all provided **integrationevents** become useless (except for internal codeunits that are used within our own apps, which receive access through the InternalVisibleTo property. For those, we can add a #pragma to skip the newly requested warning).

**Requirement**
Please add a new rule that verifies if an **IntegrationEvent** is part of an object that has been marked with **Access = Internal** and raise an Error.

**Comment by StefanMaron** [comment]
But LC0011 does not enforce access = internal. It just warns if access is not defined at all. You can set method codeunits to the default value and there should be no warning

**Comment by fvet** [comment]
I had first enabled LC0011 and made it result into a **warning**. Then - to be as safe as possible for AppSource - I put 'Access = Internal' as the default value, which is fine for most codeunits, but not for those exposing events. 

I want to prevent developers from adding new events into these Internal codeunits, which would result in the events not being available for other extensions (if they did not verify the codeunits access property)

**Comment by rvanbekkum** [comment]
I agree, you typically do not want integration events in codeunits with `Access = Internal;`, and if you really do (e.g., for testing?), then you can always suppress the warning for individual occurrences using `#pragma warning disable LC...`.

**Comment by rvanbekkum** [comment]
Seeing it got a lot of votes and I think it is useful, I implemented the rule with PR https://github.com/StefanMaron/BusinessCentral.LinterCop/pull/194.

**Comment by StefanMaron** [comment]
Sorry for the long delay of things here. I had some private things going on that took my attention. But I hope to be on track again now :)

This is now deployed to prerelease for you all to test!
